### PR TITLE
EFX tplg2

### PR DESCRIPTION
**Pull request by juimonen**

### CODE REVIEW

**Review by singalsu — APPROVED** [review]
Works well, thanks Jaska!

File: tools/topology/topology2/include/components/eqfir.conf
Comment (ranj063) [inline tools/topology/topology2/include/components/eqfir.conf:39]: why do we need this for IPC4?

File: tools/topology/topology2/include/components/eqfir/passthrough.conf
Comment (ranj063) [inline tools/topology/topology2/include/components/eqfir/passthrough.conf:1]: why is this called passthrough.conf?

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:44]: typo in name of the pipeline

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:45]: this doesnt match the attribute name, its instance isnt it?

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:113]: @juimonen @aiChaoSONG has a patch coming up for including small snippets of conf nodes without including a new file. We should wait until that is merged

File: tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf
Comment (ranj063) [inline tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf:27]: The only difference between this one and the cavs-mixin-mixout-hda.conf is what pipeline is used in the backend. Can we make this conditional based on a variable definition?

File: tools/topology/topology2/include/components/eqfir/passthrough.conf
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir/passthrough.conf:1]: so it is a flat response, does nothing - so "passthrough"

File: tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf
Comment (juimonen) [reply tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf:27]: there's also the byte control name definition... so diff is still small, but not only conf name. I can try to add some conditional logic.

File: tools/topology/topology2/include/components/eqfir.conf
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir.conf:39]: hmm ok, need to test without

File: tools/topology/topology2/include/components/eqfir/passthrough.conf
Comment (singalsu) [reply tools/topology/topology2/include/components/eqfir/passthrough.conf:1]: It's actually a blob that sets every channel filter index (preset) to -1. It instructs the EQ to pass that channel directly to output. When every channel is such the process() happens when mem copy operations for entire period.

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf
Comment (juimonen) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf]: fixed

File: tools/topology/topology2/include/components/eqfir.conf
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir.conf]: removed, works without this.

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf
Comment (juimonen) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:113]: so we actually want to have the eq blobs in separate files as they can be produced with octave tools.

File: tools/topology/topology2/development/CMakeLists.txt
Comment (ranj063) [inline tools/topology/topology2/development/CMakeLists.txt:29]: @juimonen @singalsu why is this a separate tplg? With IPC3 we had this in the regular HDA topologies isnt it?

Comment (juimonen) [reply tools/topology/topology2/development/CMakeLists.txt:29]: ok, I'll move to draft and move to generic

**Comment by juimonen** [comment]
moved to draft as I will make a change to include eq to hda-generic (it will be tested by CI). We have dependencies on kernel side
 https://github.com/thesofproject/linux/pull/4132/

**Comment by juimonen** [comment]
@ranj063 looking this again: I think PR as such is actually quite ok. You could change the "mainline" hda topology to use fir just by setting "HDA_CONFIG=fir" in cmakefile (currently it is HDA_CONFIG=mix). And I would like to have the possibility to move back and forth if needed. It is not very simple just rip off or add a component from topology (I've tried it couple of times and took me a while).

So just saying that I think this PR is good imho, we could change the mainline by follow up PR for cmakefile.

File: tools/topology/topology2/include/components/eqfir.conf
Comment (lyakh) [inline tools/topology/topology2/include/components/eqfir.conf:27]: "FIR?"

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf
Comment (juimonen) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf]: fixed

File: tools/topology/topology2/include/components/eqfir.conf
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir.conf]: fixed

**Comment by juimonen** [comment]
update: changed to generic EFX (end/post effect) pipeline, that has gain, iir and fir.

**Comment by singalsu** [comment]
Update with gain, IIR, FIR as EFX processing works here perfectly.  Patches needed in addition this are:

https://github.com/thesofproject/linux/pull/4132
https://github.com/thesofproject/sof/pull/6969
https://github.com/thesofproject/sof/pull/6992

Rimage is merged but need to make sure Zephyr build has https://github.com/thesofproject/rimage/pull/124

File: tools/topology/topology2/platform/intel/efx-default.conf
Comment (lgirdwood) [inline tools/topology/topology2/platform/intel/efx-default.conf:4]: missing newline at EOF.

Comment (juimonen) [reply tools/topology/topology2/platform/intel/efx-default.conf]: ok fixed

File: tools/topology/topology2/include/components/eqiir/passthrough.conf
Comment (lgirdwood) [reply tools/topology/topology2/include/components/eqiir/passthrough.conf:4]: This should only come from the matlab tools, I really dont want to have to use ```bytes``` in topology2. i.e. we should use 
```
file "the_eq_coeff_file.bin"
```

Comment (juimonen) [reply tools/topology/topology2/include/components/eqiir/passthrough.conf]: now updated to reflect changes in:
https://github.com/thesofproject/linux/pull/4132

File: tools/topology/topology2/include/components/eqfir/passthrough.conf
Comment (ujfalusi) [inline tools/topology/topology2/include/components/eqfir/passthrough.conf:4]: 0x38, 0x00, 0x00, 0x00.
The 0x0e is wrong.

Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir/passthrough.conf]: urgh indeed. fixed now.

**Comment by lgirdwood** [comment]
@juimonen looks like the dependencies are all merged. Still draft ?

**Review by lgirdwood — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
